### PR TITLE
Add CSRNet

### PR DESCRIPTION
**Pull request by YanLee97**

### CODE REVIEW

File: CSRNet/dataset.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        self.lines = root
+        self.transform = transform
+        self.train = train
+        self.shape = shape
+        self.seen = seen
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+    def __len__(self):
+        # 返回数据集中数据个数
```
Comment (Ldpe2G) [inline CSRNet/dataset.py:18]: 用英文写注释

File: CSRNet/README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+### predicted values on a single image:
+```
+python test.py
+```
+
+```bash
+bash infer.sh
+```
+
+### Citation
```
Comment (Ldpe2G) [inline CSRNet/README.md:54]: citation 这一节放最后

File: CSRNet/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+This dataset consists of two parts as Part A containing 482 images with highly congested scenes randomly downloaded
+from the Internet while Part B includes 716 images with relatively sparse crowd scenes taken from streets in Shang-
+hai.[ShanghaiTech ](https://oneflow-public.oss-cn-beijing.aliyuncs.com/datasets/models/CSRNet/Shanghai_dataset.rar) Dataset are provided.
+
+
+### Download Pretrain Models
+
+```bash
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/cv/CSRNet/Shanghai_BestModelA.rar
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/cv/CSRNet/Shanghai_BestModelB.rar
```
Comment (Ldpe2G) [inline CSRNet/README.md:25]: 解压的命令也加上

File: CSRNet/README.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+### Follow the val.sh to try the validation
+First, modify the address of the Pretrain Models to be used in the val.py file
+```bigquery
```
Comment (Ldpe2G) [inline CSRNet/README.md:30]: bash

File: CSRNet/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+python train.py part_A_train.json part_A_val.json 0 0
+```
+
+```bash
+bash train.sh
+```
+### predicted values on a single image:
+```
+python test.py
+```
```
Comment (Ldpe2G) [inline CSRNet/README.md:48]: 下面写了调用 infer.sh ，这里可以删了

File: CSRNet/README.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/cv/CSRNet/Shanghai_BestModelB.rar
+```
+
+### Follow the val.sh to try the validation
+First, modify the address of the Pretrain Models to be used in the val.py file
+```bigquery
+checkpoint = flow.load('checkpoint/Shanghai_BestModelA/shanghaiA_bestmodel')
+```
+```bash
+bash infer.sh
```
Comment (Ldpe2G) [inline CSRNet/README.md:34]: 这里是不是写错了，标题写 val.sh，这里调用的 infer.sh

File: CSRNet/README.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+This repo is based on: https://github.com/leeyeehoo/CSRNet-pytorch
+
+
+## Prerequisites
+
+```
+pip3 install -r requirements.txt
+```
+
+## Get started
```
Comment (Ldpe2G) [inline CSRNet/README.md:13]: Getting start

File: CSRNet/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+### Dataset
+ShanghaiTech crowd counting dataset contains 1198 an-notated images with a total amount of 330,165 persons.
+This dataset consists of two parts as Part A containing 482 images with highly congested scenes randomly downloaded
+from the Internet while Part B includes 716 images with relatively sparse crowd scenes taken from streets in Shang-
+hai.[ShanghaiTech ](https://oneflow-public.oss-cn-beijing.aliyuncs.com/datasets/models/CSRNet/Shanghai_dataset.rar) Dataset are provided.
```
Comment (Ldpe2G) [inline CSRNet/README.md:18]: 最后一句 `ShanghaiTech Dataset are provided.` 可以删了，把链接加到，这段开头的 `ShanghaiTech`  上

File: CSRNet/val.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+part_B_test = os.path.join(root,'part_B_final/test_data','images')
+path_sets = [part_A_test]
+img_paths = []
+for path in path_sets:
+    for img_path in glob.glob(os.path.join(path, '*.jpg')):
+        img_paths.append(img_path)
+
+model = CSRNet()
+model = model.to("cuda")
+checkpoint = flow.load('checkpoint/model_best')
```
Comment (Ldpe2G) [inline CSRNet/val.py:27]: 这脚本中出现的路径，都不要写死，改成脚本中参数传入

File: CSRNet/utils.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    flow.save(state['state_dict'], filename+task_id)
+    if is_best:
+        file_path= 'checkpoints/model_best'
+        del_file(file_path)
+        shutil.copytree(filename+task_id, file_path)
+
+
+def del_file(filepath):
+    """
+    删除某一目录下的所有文件或文件夹
```
Comment (Ldpe2G) [inline CSRNet/utils.py:28]: 英文

**Comment by Ldpe2G** [comment]
建议再对照 [模型验收&pr提交标准](https://github.com/Oneflow-Inc/models/issues/38)，重新整理下代码和脚本

File: CSRNet/README.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+checkpoint = flow.load('checkpoint/Shanghai_BestModelA/shanghaiA_bestmodel')
+```
+```bash
+bash infer.sh
+```
+
+### Train a model
+```
+python train.py part_A_train.json part_A_val.json 0 0
+```
```
Comment (Ldpe2G) [inline CSRNet/README.md:40]: 这句命令删掉吧，用户调 train.sh 就行了

File: CSRNet/model.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, load_weights=False):
+        super(CSRNet, self).__init__()
+        self.seen = 0
+        self.frontend_feat = [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512]
+        self.backend_feat = [512, 512, 512, 256, 128, 64]
+        self.frontend = make_layers(self.frontend_feat)
+        self.backend = make_layers(self.backend_feat, in_channels=512, dilation=True)
+        self.output_layer = nn.Conv2d(64, 1, kernel_size=1)
+
+        if not load_weights:
```
Comment (Ldpe2G) [inline CSRNet/model.py:14]: vgg 模型有用到吗，我看调用 CSRNet 的地方都没有 把 load_weights 设成 True

File: CSRNet/README.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+
+### Download Pretrain Models
+
+```bash
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/cv/CSRNet/Shanghai_BestModelA.rar
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/cv/CSRNet/Shanghai_BestModelB.rar
+```
+
+### Follow the val.sh to try the validation
```
Comment (Ldpe2G) [inline CSRNet/README.md:28]: ### Run validation

File: CSRNet/make_dateset.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
+import scipy.io as io
+import os
+import glob
+from matplotlib import pyplot as plt
+from scipy.ndimage.filters import gaussian_filter
+import scipy
+from image import *
+import scipy.spatial
+
```
Comment (Ldpe2G) [inline CSRNet/make_dateset.py:9]: 这头文件引用顺序看着也很乱，重新真理一下吧

File: CSRNet/make_dateset.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            hf['density'] = k
+
+'''plt.imshow(Image.open(img_paths[0]))
+
+gt_file = h5py.File(img_paths[0].replace('.jpg', '.h5').replace('images', 'ground_truth'), 'r')
+groundtruth = np.asarray(gt_file['density'])
+plt.imshow(groundtruth, cmap=CM.jet)
+
+np.sum(groundtruth)'''
+
```
Comment (Ldpe2G) [inline CSRNet/make_dateset.py:70]: 删掉无用的注释

File: CSRNet/model.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            mod = vgg16(pretrained=True)
+            pretrain_models = flow.load('vgg_imagenet_pretrain_model/vgg16_oneflow_model')
+            mod.load_state_dict(pretrain_models)
+            self._initialize_weights()
+            #print(self.frontend.state_dict().items())
```
Comment (Ldpe2G) [inline CSRNet/model.py:19]: 删

File: CSRNet/test.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+temp = h5py.File('dataset/part_A_final/test_data/ground_truth/IMG_100.h5', 'r')
+temp_1 = np.asarray(temp['density'])
+plt.title("Original Count")
+plt.imshow(temp_1,cmap = c.jet)
+print("Original Count : ",int(np.sum(temp_1)) + 1)
+plt.show()
+print("Original Image")
+plt.title("Original Image")
+plt.imshow(plt.imread('dataset/part_A_final/test_data/images/IMG_100.jpg'))
+plt.show()
```
Comment (Ldpe2G) [inline CSRNet/test.py:36]: 同样这脚本中出现的路径都不要写死，改成 infer.sh 脚本中传入

File: CSRNet/train.py [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+from model import CSRNet
+from utils import save_checkpoint
+import argparse
+import json
+import time
+import oneflow as flow
+import oneflow.nn as nn
+import transforms.spatial_transforms as ST
+
+parser = argparse.ArgumentParser(description='PyTorch CSRNet')
```
Comment (Ldpe2G) [inline CSRNet/train.py:12]: Oneflow CSRNet

File: CSRNet/train.py [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+
+
+    if args.pre:
+        if os.path.isfile(args.pre):
+            print("=> loading checkpoint '{}'".format(args.pre))
+            checkpoint = flow.load(args.pre)
+            args.start_epoch = checkpoint['epoch']
+            best_prec1 = checkpoint['best_prec1']
+            model.load_state_dict(checkpoint['state_dict'])
+            #7.21尝试oneflow版本能否加载
```
Comment (Ldpe2G) [inline CSRNet/train.py:66]: 这个是 todo 还是？

File: CSRNet/train.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+        is_best = prec1 < best_prec1
+        best_prec1 = min(prec1, best_prec1)
+        print(' * best MAE {mae:.3f} '
+              .format(mae=best_prec1))
+        save_checkpoint({
+            'epoch': epoch + 1,
+            'arch': args.pre,
+            'state_dict': model.state_dict(),
+            'best_prec1': best_prec1,
+            #'optimizer': optimizer.state_dict(),
```
Comment (Ldpe2G) [inline CSRNet/train.py:90]: 删掉无用注释

File: CSRNet/train.py [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+    model.eval()
+    mae = 0
+    for i, (img, target) in enumerate(test_loader):
+        img = flow.Tensor(img, dtype=flow.float32, device="cuda")
+        with flow.no_grad():
+            output = model(img).to("cuda")
+
+
+       # output = model(img).to("cuda")
+       # output = flow.Tensor(output, device="cuda")
```
Comment (Ldpe2G) [inline CSRNet/train.py:157]: 删掉无用注释

File: CSRNet/transforms/temporal_transforms.py [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+#             out = frame_indices[0:4:1]
+#         else:
+#             out = frame_indices[0:4]
+#             while len(out) < 4:
+#                 for index in out:
+#                     if len(out) >= 4:
+#                         break
+#                     out.append(index)
+
+#         return out
```
Comment (Ldpe2G) [inline CSRNet/transforms/temporal_transforms.py:172]: 删掉无用注释

File: CSRNet/utils.py [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    with h5py.File(fname, 'r') as h5f:
+        for k, v in net.state_dict().items():        
+            param = flow.Tensor(np.asarray(h5f[k]))
+            v.copy_(param)
+def save_checkpoint(state, is_best,task_id, filename='checkpoints/'):
+
+    del_file(filename+str(int(task_id) - 1))
+    flow.save(state['state_dict'], filename+task_id)
+    if is_best:
+        file_path= 'checkpoints/model_best'
```
Comment (Ldpe2G) [inline CSRNet/utils.py:21]: 不要写死路径

File: CSRNet/model.py (hunk 12 above)
Comment (YanLee97) [reply CSRNet/model.py:14]: 这边源代码是False

File: CSRNet/utils.py (hunk 10 above)
Comment (YanLee97) [reply CSRNet/utils.py]: 已经修改

File: CSRNet/infer.py [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+def main():
+    transform = ST.Compose([
+        ST.ToNumpyForVal(), ST.Normalize(mean=[0.485, 0.456, 0.406],
+                                         std=[0.229, 0.224, 0.225]),
+    ])
+    global args
+    args = parser.parse_args()
+    model = CSRNet()
+    model = model.to("cuda")
+    # checkpoint = flow.load('checkpoint/Shanghai_BestModelA/shanghaiA_bestmodel')
```
Comment (Ldpe2G) [inline CSRNet/infer.py:26]: 无用代码删除

**Review by Ldpe2G — APPROVED** [review]
